pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate=true
+package v1beta1
+
+import (
+	core_v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeSnapshot is a user's request for taking a point-in-time snapshot of a PersistentVolumeClaim.
+// Upon successful creation of a snapshot by the underlying storage system, it is bound to a
+// corresponding VolumeSnapshotContent.
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec defines the desired characteristics of a snapshot requested by a user.
+	// More info: https://kubernetes.io/docs/concepts/storage/volume-snapshots#volumesnapshots
+	// Required.
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status represents the current information of a snapshot.
+	// NOTE: status can be modified by sources other than system controllers,
+	// and must not be depended upon for accuracy.
+	// Controllers should only use information from the VolumeSnapshotContent object
+	// after verifying that the binding is accurate and complete.
+	// +optional
+	Status VolumeSnapshotStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
since it is optional, should it be a pointer? Although all other API does also use struct instead of pointer. Not sure why...

Reply:
That's a good question. I spoke with liggitt. He agreed and recommended we go with optional pointers.  And default status to `{}`

Reply:
We should also make sure that we're specifying status as a subresource

Reply:
status has been defined as subresource.
The existing pattern in K8s seems always put it as a value instead of a pointer, and all the fields in status are optional anyways. Users are not capable of updating status via kubectl. is it really necessary?

Reply:
The [api conventions doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) says that all optional values should be pointers, so we should follow that going forward.  The fact that it's not a pointer in K8s is mainly due to legacy before the current guidelines were developed.

Reply:
changed to a pointer along with controller/test code changes.